modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverSetUtils.java

```diff
@@ -109,107 +127,170 @@ private static Map<String, ValueResolver> getParameterGroupValueResolvers(Parame
       throws MuleException {
-    Class expectedType = parameterType.getDeclaringClass().orElse(Collection.class);
-    List<Type> parameterTypeCollectionGenerics = parameterType.getSuperTypeGenerics(Collection.class);
-    if (parameterTypeCollectionGenerics.size() == 1) {
+    Optional<Class<Object>> expectedType = getType(arrayType);
+    if (expectedType.isPresent()) {
+      Class type = expectedType.get();
       List<ValueResolver<Object>> itemsResolvers = new ArrayList<>();
       for (Object collectionItem : collection) {
         itemsResolvers
-            .add(getParameterValueResolver(parameterTypeCollectionGenerics.get(0), collectionItem, reflectionCache, muleContext));
+            .add(getParameterValueResolver(parameterName, arrayType.getType(), collectionItem, emptySet(), reflectionCache,
+                                           muleContext, valueResolverFactory));
       }
-
-      return CollectionValueResolver.of(expectedType, itemsResolvers);
+      return CollectionValueResolver.of(type, itemsResolvers);
     } else {
       return new StaticValueResolver(collection);
     }
   }
 
-  private static ValueResolver getParameterValueResolverForMap(Type parameterType, Map<Object, Object> map,
+  private static ValueResolver getParameterValueResolverForMap(String parameterName, ObjectType type, Map<Object, Object> map,
                                                                ReflectionCache reflectionCache,
-                                                               MuleContext muleContext)
+                                                               MuleContext muleContext, ValueResolverFactory valueResolverFactory)
       throws MuleException {
-    Class expectedType = parameterType.getDeclaringClass().orElse(Map.class);
-    List<Type> parameterTypeCollectionGenerics = parameterType.getSuperTypeGenerics(Map.class);
-    if (parameterTypeCollectionGenerics.size() == 2) {
-      List<ValueResolver<Object>> keyResolvers = new ArrayList<>();
-      List<ValueResolver<Object>> valueResolvers = new ArrayList<>();
-      for (Map.Entry<Object, Object> mapEntry : map.entrySet()) {
-        keyResolvers.add(getParameterValueResolver(parameterTypeCollectionGenerics.get(0), mapEntry.getKey(), reflectionCache,
-                                                   muleContext));
-        valueResolvers.add(getParameterValueResolver(parameterTypeCollectionGenerics.get(1), mapEntry.getValue(), reflectionCache,
-                                                     muleContext));
-      }
+    Optional<Class<Object>> mapClassOptional = getType(type);
+    Class mapClass = mapClassOptional.get();
 
-      return MapValueResolver.of(expectedType, keyResolvers, valueResolvers, reflectionCache, muleContext);
-    } else {
-      return new StaticValueResolver(map);
+    MetadataType valueType = type.getOpenRestriction().orElse(null);
+    Function<Object, ValueResolver<Object>> valueValueResolverFunction = valueType != null
```

Comment:
rewrite this code for clarity. The ternary + the lambda makes this hard to read and debug

Reply:
👌